hazelcast/src/main/java/com/hazelcast/map/impl/query/PartitionScanRunner.java

```diff
@@ -94,20 +97,24 @@ public void run(String mapName, Predicate predicate, int partitionId, Result res
 
             @Override
             public void accept(Data key, Record record) {
-                Metadata metadata = PartitionScanRunner.this.getMetadataFromRecord(recordStore, key, record);
-                Object value = PartitionScanRunner.this.toData(useCachedValues
-                        ? Records.getValueOrCachedValue(record, serializationService)
-                        : record.getValue());
+                Object value = useCachedValues ? getValueOrCachedValue(record, ss) : record.getValue();
+                // TODO how can a value be null?
                 if (value == null) {
                     return;
                 }
 
-                queryEntry.init(serializationService, key, value, extractors);
-                queryEntry.setMetadata(metadata);
+                // always copy key&value to heap
+                key = toHeapData(key);
```

Comment:
Why not to check `nativeMemory` flag for the key as well?

Reply:
value can be in Data or in Object type, depending on the configured in-memory-format but key can only be in Data type.
`nativeMemory` flag is needed to prevent unneeded conversion of value to Data from an Object type. But this is not the case for keys since no unneeded conversion can be done with keys, they are always in Data format.